mlflow/types/utils.py

```diff
@@ -133,7 +153,9 @@ def _infer_array_datatype(data: Union[list, np.ndarray]) -> Optional[Array]:
         ["a", "b"] => Array(string)
         ["a", None] => Array(string)
         [["a", "b"], []] => Array(Array(string))
+        [["a", "b"], None] => Array(Array(string))
```

Comment:
Possibly follow-up: What about `["a", 1, 0.1]`? Shouldn't we infer `Array(Any)` for this kind of case?

Reply:
I'm a bit hesitate for this case, initially I think we should support Union type, but considering we want to support pydantic models in the future then I don't think we should spend more time supporting Union types.

Reply:
>  I think we should wait until such requirements come up

We've already heard it a few times e.g. multi-modal model inputs. By definition `Any` type does not enforce any restriction for the data type, which is not related to `Union` type. If you consider python's type system, `list[Any]` works for heterogenous lists. I don't think we should add such restriction to `Any` type for saving effort.

> considering we want to support pydantic models

Pydantic model will only used for GenAI flavors right? IIUC `Any` type is not limited for GenAI.

Reply:
I think `Any` type should be limited to GenAI flavors;
For multi-modal model inputs seem it fits more into `Union` type.

Reply:
@BenWilson2 @harupy WDYT?

Reply:
I don't want us to infer AnyType from [1, 1.0], as I think it's not safe for traditional ML models, but users can explicitly define it if they want to

Reply:
Ok, what kind of error message we emit from schema inference for those input? Can we at least update it to suggest `Any` type?

Reply:
Yep that makes sense, currently it's `Expected all values in list to be of same type`